File: delete_leaves.py
class Node(object):
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

    def __repr__(self):
        result = str(self.value)
        if self.left:
            result += str(self.left)
        if self.right:
            result += str(self.right)
        return result

class Solution():
    def __init__(self):
        self.name = 'S->13'

    def __DeleteLeavesHelper(self, root, target):
        if not root:
            return None
        root.left = self.__DeleteLeavesHelper(root.left, target)
        root.right = self.__DeleteLeavesHelper(root.right, target)
        if root.value == target and not root.left and not root.right:
            return None

        return root

    # Recursive method
    # Time complexity: O(n)
    # Space complexity: O(n)
    def DeleteLeavesRecursive(self, root, target):
        return self.__DeleteLeavesHelper(root, target)

    def DeleteLeaves(self, root, target):
        stack = []
        parent = {}
        cur = root
        pre = None
        while True:
            if cur:
                if cur.right:
                    stack.append(cur.right)
                    if cur.right not in parent:
                        parent[cur.right] = [cur, 'right']
                stack.append(cur)
                if pre != None and cur not in parent:
                    parent[cur] = [pre, 'left']
                pre = cur
                cur = cur.left
                print(parent)
            else:
                if stack:
                    node = stack.pop()
                    if stack and node.right == stack[-1]:
                        stack.pop()
                        cur = node.right
                        stack.append(node)
                    elif node.value == target and not node.left and not node.right:
                        if parent[node][1] == 'right':
                            parent[node][0].right = None
                        if parent[node][1] == 'left':
                            parent[node][0].left = None
                else:
                    break

        return root

#     1
#    / \
#   2   1
#  /   /
# 2   1
s = Solution()
n1 = Node(1)
n1.left = Node(1)
n1.right = Node(1)
print(n1)
print(s.DeleteLeaves(n1, 1))
print(s.DeleteLeavesRecursive(n1, 1))

File: Data_Structure/Trees/delete_leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    typedef std::pair<TreeNode*, string> pp;
    
    // Time complexity: O(n)
    // Space complexity: O(n)
    TreeNode* removeLeafNodes(TreeNode* root, int target) {
        unordered_map<TreeNode*, pp> parent;
        stack<TreeNode*> s;
        TreeNode* cur = root;
        TreeNode* pre = nullptr;
        while (true) {
            if (cur) {
                if (cur->right) {
                    s.push(cur->right);
                    if (parent.find(cur->right) == parent.end()) {
                        parent[cur->right] = {cur, "right"};    
                    }
                }
                s.push(cur);
                if (pre && parent.find(cur) == parent.end()) {
                    parent[cur] = {pre, "left"};
                }
                pre = cur;
                cur = cur->left;
            }
            else {
                if (!s.empty()) {
                    TreeNode* node = s.top();
                    s.pop();
                    if (!s.empty() && node->right &&node->right == s.top()) {
                        s.pop();
                        cur = node->right;
                        s.push(node);
                    }
                    else {
                        if (node->val == target && !node->left && !node->right) {
                            if (node == root) {
                                return nullptr;
                            }
                            if (parent[node].second == "left") {
                                parent[node].first->left = nullptr;
                            }
                            if (parent[node].second == "right") {
                                parent[node].first->right = nullptr;
                            }
                        }
                    }
                }
                else {
                    break;
                }
            }
        }
        return root;
    }
};
